src/Abstract/Syntax.hs

```diff
@@ -1,283 +1,437 @@
+instance Show1 Variable where liftShowsPrec = genericLiftShowsPrec
+instance Eq1 Variable where liftEq _ (Variable name1) (Variable name2) = name1 == name2
+instance Ord1 Variable where liftCompare _ (Variable name1) (Variable name2) = compare name1 name2
+
+instance (Monad m, MonadFail m, MonadAddress l m, MonadStore l (Value s l) m, MonadEnv l (Value s l) m) => Eval (Value s l) m s Variable where
+  evaluate _ (Variable x) = do
+    env <- askEnv
+    maybe (fail ("free variable: " ++ x)) deref (envLookup x (env :: Environment l (Value s l)))
+
+instance (Alternative m, MonadFail m, MonadStore Monovariant Type m, MonadEnv Monovariant Type m) => Eval Type m s Variable where
+  evaluate _ (Variable x) = do
+    env <- askEnv
+    maybe (fail ("free type: " ++ x)) deref (envLookup x (env :: Environment Monovariant Type))
+
+instance FreeVariables1 Variable where
+  liftFreeVariables _ (Variable name) = point name
+
+instance (Monad m, Eval v m s Variable) => EvalCollect l v m s Variable
+
+
+-- Primitives
+newtype Primitive a = Primitive Prim deriving (Eq, Ord, Show, Functor, Foldable, Generic1, FreeVariables1)
+instance Show1 Primitive where liftShowsPrec = genericLiftShowsPrec
+instance Eq1 Primitive where liftEq _ (Primitive a1) (Primitive a2) = a1 == a2
+instance Ord1 Primitive where liftCompare _ (Primitive a1) (Primitive a2) = compare a1 a2
+
+instance Monad m => Eval (Value s l) m s Primitive where
+  evaluate _ (Primitive x) = return (I x)
+
+instance Monad m => Eval Type m s Primitive where
+  evaluate _ (Primitive (PInt _)) = return Int
+  evaluate _ (Primitive (PBool _)) = return Bool
+
+instance (Monad m, Eval v m s Primitive) => EvalCollect l v m s Primitive
+
+-- Lambdas
+data Lambda a = Lambda Name a deriving (Eq, Ord, Show, Functor, Foldable, Generic1)
+instance Show1 Lambda where liftShowsPrec = genericLiftShowsPrec
+instance Eq1 Lambda where liftEq comp (Lambda name1 body1) (Lambda name2 body2) = name1 == name2 && comp body1 body2
+instance Ord1 Lambda where liftCompare comp (Lambda name1 body1) (Lambda name2 body2) = compare name1 name2 `mappend` comp body1 body2
+
+instance (Monad m, MonadEnv l (Value s l) m) => Eval (Value s l) m s Lambda where
+  evaluate _ (Lambda name body) = do
+    env <- askEnv
+    return (Closure name body (env :: Environment l (Value s l)))
+
+instance (MonadStore Monovariant Type m, MonadEnv Monovariant Type m, MonadFail m, Semigroup (Cell Monovariant Type), MonadFresh m, Alternative m) => Eval Type m s Lambda where
+  evaluate ev (Lambda name body) = do
+    a <- alloc name
+    tvar <- fresh
+    assign a (TVar tvar)
+    outTy <- localEnv (envInsert name (a :: Address Monovariant Type)) (ev body)
+    return (TVar tvar :-> outTy)
+
+instance FreeVariables1 Lambda where
+  liftFreeVariables f (Lambda name body) = delete name (f body)
+
+instance (Monad m, Eval v m s Lambda) => EvalCollect l v m s Lambda
+
+-- Recursive lambdas
```

Comment:
`Rec` is a binder, but not actually a lambda—it’s more like `letrec` in ML (or just `let` in Haskell), i.e. it introduces a binding which is in scope in its definition. Defining a recursive function requires both a `Rec` and a `Lambda`; if you only have a `Rec`, you can define recursively-constructed values, but not functions.